README.md
[![Gem Version](https://badge.fury.io/rb/react_rails_modal.svg)](https://badge.fury.io/rb/react_rails_modal)

# ReactRailsModal

Accessible modal dialog component on react-rails.

## Pre Installation

Before using react_rails_modal, install `react-rails`.

Add `react-rails` to your gemfile:

```ruby
gem 'react-rails'
```

And install:

    $ bundle install

Next, run the installation script:

    $ rails g react:install

## Installation

Add `react_rails_modal` to your gemfile:

```ruby
gem 'react_rails_modal'
```

And install:

    $ bundle install

Next, run the installation script:

    $ rails g react_rails_modal:install

This will place the following in your `application.js`:

```js
//= require react_rails_modal
```

## Usage

```xml
<ModalComponent
  isOpen={bool}
  onCloseRequest={requestCloseFunction}
  style={customStyle}
>
  <h1>Modal Component Content</h1>
  <p>Text</p>
</ModalComponent>
```

## Examples
### Use styles to style

In app/controller/views/:

```html
<%= react_component('Hello') %>
```

Styles are passed as an object with 2 keys, 'overlay' and 'modal'.

In app/assets/javascripts/components/:

```jsx
class Hello extends React.Component {
  constructor(props) {
    super(props);
    this.closeModal = this.closeModal.bind(this);
    this.openModal = this.openModal.bind(this);
    this.state = { isModalOpen: false };
  }

  closeModal() {
    this.setState({ isModalOpen: false });
  }

  openModal() {
    this.setState({ isModalOpen: true });
  }

  render () {
    return (
      <div>
        <button onClick={this.openModal}>Open Modal</button>
        <ModalComponent
          isOpen={this.state.isModalOpen}
          onCloseRequest={this.closeModal}
          style={Hello.styles}
        >
          <h1>Hello world!</h1>
          <button onClick={this.closeModal}>Close</button>
        </ModalComponent>
      </div>
    );
  }
}

Hello.styles = {
  overlay: {
    backgroundColor: 'rgba(0,0,0,0.75)',
    bottom: '0',
    height: '100%',
    position: 'fixed',
    top: '0',
    left: '0',
    right: '0',
    width: '100%',
    zIndex: '1',
  },
  modal: {
    background: '#fff',
    left: '5%',
    padding: '15px 30px',
    position: 'relative',
    top: '10%',
    width: '80%',
    zIndex: '2',
  }
}
```

### Use css classes to style

If you want to use css classes, you can pass overlayClass and modalClass props to the Modal.

```jsx
class HelloTwo extends React.Component {
  constructor(props) {
    super(props);
    this.closeModal = this.closeModal.bind(this);
    this.openModal = this.openModal.bind(this);
    this.state = { isModalOpen: false };
  }

  closeModal() {
    this.setState({ isModalOpen: false });
  }

  openModal() {
    this.setState({ isModalOpen: true });
  }

  render () {
    return (
      <div>
        <button onClick={this.openModal}>Open Modal</button>
        <ModalComponent
          isOpen={this.state.isModalOpen}
          onCloseRequest={this.closeModal}
          overlayClass='overlayClass'
          modalClass='modalClass'
        >
          <h1>Hello world!!</h1>
          <button onClick={this.closeModal}>Close</button>
        </ModalComponent>
      </div>
    );
  }
}
```

In app/assets/stylesheets/:

```css
.overlayClass {
  background-color: rgba(0,0,0,0.7);
  bottom: 0;
  height: 100%;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  z-index: 1;
}

.modalClass {
  background: #fff;
  left: 5%;
  padding: 15px;
  position: fixed;
  top: 10%;
  width: 90%;
  z-index: 2;
}
```

## Development
* Update React assets with `rake react_rails_modal:build`
* Run tests with `rake spec`


CHANGELOG.md
# react_rails_modal
## v0.1.3 (December 9, 2016)
[fixed] Fix Usage of README

## v0.1.2 (December 7, 2016)
[fixed] Remove unnecessary bundle.js file

## v0.1.1 (December 7, 2016)
[fixed] Remove unnecessary AssetVariant class

## v0.1.0 (December 6, 2016)


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "react_rails_modal"


spec/react_rails_modal_spec.rb
require "spec_helper"

describe ReactRailsModal do
  it "has a version number" do
    expect(ReactRailsModal::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/react_rails_modal/version_spec.rb
require 'spec_helper'

module ReactRailsModal
  describe 'version' do
    it 'has a version number' do
      expect(ReactRailsModal::VERSION).not_to be nil
    end
  end
end


react_builds/app.bundle.js
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};

/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {

/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;

/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			exports: {},
/******/ 			id: moduleId,
/******/ 			loaded: false
/******/ 		};

/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

/******/ 		// Flag the module as loaded
/******/ 		module.loaded = true;

/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}


/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;

/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;

/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";

/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ function(module, exports) {

	"use strict";

	var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

	function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

	function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

	function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

	var ModalComponent = function (_React$Component) {
	  _inherits(ModalComponent, _React$Component);

	  function ModalComponent() {
	    _classCallCheck(this, ModalComponent);

	    return _possibleConstructorReturn(this, (ModalComponent.__proto__ || Object.getPrototypeOf(ModalComponent)).apply(this, arguments));
	  }

	  _createClass(ModalComponent, [{
	    key: "render",
	    value: function render() {
	      return React.createElement(
	        "div",
	        null,
	        "hogehoge"
	      );
	    }
	  }]);

	  return ModalComponent;
	}(React.Component);

	// export { ModalComponent };
	// module.exports = ModalComponent;

/***/ }
/******/ ]);

lib/generators/react_rails_modal/install_generator.rb
module ReactRailsModal
  module Generators
    class InstallGenerator < ::Rails::Generators::Base
      desc 'Install react_rails_modal'

      def inject_react_rails_modal
        require_react_rails_modal = "//= require react_rails_modal\n"

        if manifest.exist?
          manifest_contents = File.read(manifest)

          if match = manifest_contents.match(/\/\/=\s+require_tree[^\n]*/)
            inject_into_file manifest, require_react_rails_modal, { before: match[0] }
          else
            append_file manifest, require_react_rails_modal
            inject_react_rails
          end
        else
          create_file manifest, require_react_rails_modal
          inject_react_rails
        end
      end

      private

      def manifest
        Pathname.new(destination_root).join('app/assets/javascripts', 'application.js')
      end

      def inject_react_rails
        inject_into_file manifest, "//= require react\n", {before: "//= require react_rails_modal\n"}
        inject_into_file manifest, "//= require components\n", {after: "//= require react\n"}
        inject_into_file manifest, "//= require react_ujs\n", {after: "//= require react\n"}
      end
    end
  end
end


lib/assets/javascripts/react_rails_modal.js
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};

/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {

/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;

/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			exports: {},
/******/ 			id: moduleId,
/******/ 			loaded: false
/******/ 		};

/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

/******/ 		// Flag the module as loaded
/******/ 		module.loaded = true;

/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}


/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;

/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;

/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";

/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ function(module, exports) {

	'use strict';

	Object.defineProperty(exports, "__esModule", {
	  value: true
	});

	var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

	function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

	function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

	function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

	var ModalComponent = function (_React$Component) {
	  _inherits(ModalComponent, _React$Component);

	  function ModalComponent(props) {
	    _classCallCheck(this, ModalComponent);

	    var _this = _possibleConstructorReturn(this, (ModalComponent.__proto__ || Object.getPrototypeOf(ModalComponent)).call(this, props));

	    _this.state = { isOpen: _this.props.isOpen };
	    return _this;
	  }

	  _createClass(ModalComponent, [{
	    key: 'componentWillReceiveProps',
	    value: function componentWillReceiveProps(nextProps) {
	      this.setState({ isOpen: nextProps.isOpen });
	    }
	  }, {
	    key: 'render',
	    value: function render() {
	      if (!this.state.isOpen) {
	        return null;
	      }

	      var overlay = this.props.style && this.props.style.overlay ? this.props.style.overlay : ModalComponent.defaultStyles.overlay;
	      var overlayClass = this.props.overlayClass ? {} : overlay;
	      var modal = this.props.style && this.props.style.modal ? this.props.style.modal : ModalComponent.defaultStyles.modal;
	      var modalClass = this.props.modalClass ? {} : modal;

	      return React.createElement(
	        'div',
	        null,
	        React.createElement(
	          'div',
	          { style: overlayClass, onClick: this.props.onCloseRequest, className: this.props.overlayClass },
	          React.createElement(
	            'div',
	            { style: modalClass, className: this.props.modalClass },
	            this.props.children
	          )
	        )
	      );
	    }
	  }]);

	  return ModalComponent;
	}(React.Component);

	ModalComponent.propTypes = {
	  isOpen: React.PropTypes.bool.isRequired,
	  onCloseRequest: React.PropTypes.func,
	  style: React.PropTypes.objectOf(React.PropTypes.object),
	  overlayClassName: React.PropTypes.string,
	  modalClassName: React.PropTypes.string
	};

	ModalComponent.defaultStyles = {
	  overlay: {
	    backgroundColor: 'rgba(0,0,0,0.75)',
	    height: '100%',
	    left: 0,
	    position: 'fixed',
	    top: 0,
	    width: '100%',
	    zIndex: 1
	  },
	  modal: {
	    background: '#fff',
	    borderRadius: '5px',
	    boxShadow: '0 0 10 #000',
	    display: 'table',
	    left: '5%',
	    maxHeight: '80%',
	    maxWidth: '80%',
	    overflowY: 'scroll',
	    padding: '20px 50px',
	    position: 'fixed',
	    top: '10%',
	    width: '80%',
	    zIndex: 2
	  }
	};

	exports.default = window.ModalComponent = ModalComponent;

/***/ }
/******/ ]);